viadot/sources/bigquery.py

```diff
@@ -0,0 +1,206 @@
+import pandas as pd
+import pandas_gbq
+from google.oauth2 import service_account
+from pydantic import BaseModel
+
+from typing import List
+
+from ..config import get_source_credentials
+from ..exceptions import CredentialError, DBDataAccessError
+from .base import Source
+from ..utils import add_viadot_metadata_columns
+
+
+class BigQueryCredentials(BaseModel):
+    type: str
+    project_id: str
+    private_key_id: str
+    private_key: str
+    client_email: str
+    client_id: str
+    auth_uri: str
+    token_uri: str
+    auth_provider_x509_cert_url: str
+    client_x509_cert_url: str
+
+
+class BigQuery(Source):
+    """
+    Description:
+    Class to connect with Bigquery project and SQL tables.
+
+    Note that credentials used for authentication can be generated only for User
+    Principal who has access to specific BigQuery project.
+
+    Documentation for this API is located at: https://cloud.google.com/bigquery/docs.
+
+    Args:
+    config_key (str, optional): The key in the viadot config holding credentials.
+        Defaults to None.
+    credentials Dict[str, Any], optional: Credentials for API connection configuration
+        (`api_key` and `user`).
+
+    Raises:
+        CredentialError: In case the credentials cannot be found/ inncorect credentials.
+    """
+
+    def __init__(
+        self,
+        config_key: str = None,
+        credentials: BigQueryCredentials = None,
+        *args,
+        **kwargs,
+    ):
+        credentials = credentials or get_source_credentials(config_key)
+
+        if credentials == None:
+            raise CredentialError("Credentials not found.")
+        
+        if not (
+            credentials.get("type")
+            and credentials.get("project_id")
+            and credentials.get("private_key_id")
+            and credentials.get("private_key")
+            and credentials.get("client_email")
+            and credentials.get("client_id")
+            and credentials.get("auth_uri")
+            and credentials.get("token_uri")
+            and credentials.get("auth_provider_x509_cert_url")
+            and credentials.get("client_x509_cert_url")
```

Comment:
I guess it is supposed to check that if at least one of the properties is not available, an error should be raised.

I have two comments:
* Perhaps you meant the `or` instruction rather than the `and` instruction.
* Validation probably needs to be more complex than just checking if properties are present and there is [Pydentic](https://docs.pydantic.dev/latest/), which allows you to describe structures by enumerating and describing the incoming data for further validation, intra-communication, test writing, use in controllers and aggregators.